File: backend/src/services/apriso/AprisoService.ts
import axios, { AxiosInstance } from 'axios';
import { logger } from '../../utils/logger';
import { config } from '../../config';
import { db } from '../../db/database';
import type { QueuedEvent } from '../queue/EventQueueService';

interface SettingRow {
  value: string;
}

function getEffectiveConfig(): { baseUrl: string; apiKey: string; username: string; password: string } {
  const getSetting = (key: string): string | undefined => {
    const row = db.prepare('SELECT value FROM settings WHERE key = ?').get(key) as SettingRow | undefined;
    return row?.value;
  };

  return {
    baseUrl: getSetting('aprisoBaseUrl') || config.apriso.baseUrl,
    apiKey: getSetting('aprisoApiKey') || config.apriso.apiKey,
    username: getSetting('aprisoUsername') || config.apriso.username,
    password: getSetting('aprisoPassword') || config.apriso.password,
  };
}

function isStubMode(baseUrl: string): boolean {
  return !baseUrl || baseUrl.includes('mock') || baseUrl.includes('localhost') || !baseUrl.startsWith('http');
}

function createAxiosInstance(cfg: ReturnType<typeof getEffectiveConfig>): AxiosInstance {
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };

  if (cfg.apiKey) {
    headers['X-API-Key'] = cfg.apiKey;
  }

  return axios.create({
    baseURL: cfg.baseUrl,
    headers,
    timeout: 15000,
    auth: cfg.username ? { username: cfg.username, password: cfg.password } : undefined,
  });
}

export class AprisoService {
  async publishMachineStatus(event: QueuedEvent): Promise<void> {
    const cfg = getEffectiveConfig();

    if (isStubMode(cfg.baseUrl)) {
      logger.info(`STUB: would publish MACHINE_STATUS to Apriso for equipment=${event.equipment_name} id=${event.id}`);
      await new Promise((resolve) => setTimeout(resolve, 200));
      return;
    }

    const client = createAxiosInstance(cfg);
    const response = await client.post('/api/v1/machine-status', {
      eventId: event.id,
      equipmentId: event.equipment_id,
      equipmentName: event.equipment_name,
      timestamp: new Date().toISOString(),
      ...event.payload,
    });

    logger.info(`Apriso MACHINE_STATUS published: event=${event.id} status=${response.status}`);
  }

  async publishCounter(event: QueuedEvent): Promise<void> {
    const cfg = getEffectiveConfig();

    if (isStubMode(cfg.baseUrl)) {
      logger.info(`STUB: would publish COUNTER to Apriso for equipment=${event.equipment_name} id=${event.id}`);
      await new Promise((resolve) => setTimeout(resolve, 200));
      return;
    }

    const client = createAxiosInstance(cfg);
    const response = await client.post('/api/v1/counter', {
      eventId: event.id,
      equipmentId: event.equipment_id,
      equipmentName: event.equipment_name,
      timestamp: new Date().toISOString(),
      ...event.payload,
    });

    logger.info(`Apriso COUNTER published: event=${event.id} status=${response.status}`);
  }

  async testConnection(baseUrl: string, username?: string, password?: string, apiKey?: string): Promise<{ success: boolean; message: string }> {
    if (isStubMode(baseUrl)) {
      return { success: true, message: 'Apriso is in stub/mock mode. Connection test simulated successfully.' };
    }

    try {
      const headers: Record<string, string> = { 'Content-Type': 'application/json' };
      if (apiKey) headers['X-API-Key'] = apiKey;

      const client = axios.create({
        baseURL: baseUrl,
        headers,
        timeout: 10000,
        auth: username ? { username, password: password || '' } : undefined,
      });

      await client.get('/api/v1/health');
      return { success: true, message: 'Apriso connection successful' };
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      return { success: false, message: `Apriso connection failed: ${message}` };
    }
  }
}

export const aprisoService = new AprisoService();

File: backend/services/apriso/apriso_service.py
import asyncio
from typing import Any

import httpx

from config import config
from db import get_db
from utils.logger import logger


def _get_effective_config() -> dict:
    db = get_db()

    def _get(key: str) -> str:
        row = db.execute("SELECT value FROM settings WHERE key = ?", (key,)).fetchone()
        return row["value"] if row else ""

    return {
        "base_url": _get("aprisoBaseUrl") or config.apriso.base_url,
        "api_key": _get("aprisoApiKey") or config.apriso.api_key,
        "username": _get("aprisoUsername") or config.apriso.username,
        "password": _get("aprisoPassword") or config.apriso.password,
    }


def _is_stub(base_url: str) -> bool:
    return not base_url or "mock" in base_url or "localhost" in base_url or not base_url.startswith("http")


def _build_client(cfg: dict) -> httpx.AsyncClient:
    headers = {"Content-Type": "application/json"}
    if cfg["api_key"]:
        headers["X-API-Key"] = cfg["api_key"]
    auth = (cfg["username"], cfg["password"]) if cfg["username"] else None
    return httpx.AsyncClient(base_url=cfg["base_url"], headers=headers, auth=auth, timeout=15.0)


class AprisoService:
    async def publish_machine_status(self, event: dict) -> None:
        cfg = _get_effective_config()
        if _is_stub(cfg["base_url"]):
            logger.info(
                f"STUB: would publish MACHINE_STATUS to Apriso "
                f"equipment={event['equipment_name']} id={event['id']}"
            )
            await asyncio.sleep(0.2)
            return

        async with _build_client(cfg) as client:
            from datetime import datetime, timezone
            response = await client.post(
                "/api/v1/machine-status",
                json={
                    "eventId": event["id"],
                    "equipmentId": event["equipment_id"],
                    "equipmentName": event["equipment_name"],
                    "timestamp": datetime.now(timezone.utc).isoformat(),
                    **event["payload"],
                },
            )
            response.raise_for_status()
            logger.info(f"Apriso MACHINE_STATUS published: event={event['id']} status={response.status_code}")

    async def publish_counter(self, event: dict) -> None:
        cfg = _get_effective_config()
        if _is_stub(cfg["base_url"]):
            logger.info(
                f"STUB: would publish COUNTER to Apriso "
                f"equipment={event['equipment_name']} id={event['id']}"
            )
            await asyncio.sleep(0.2)
            return

        async with _build_client(cfg) as client:
            from datetime import datetime, timezone
            response = await client.post(
                "/api/v1/counter",
                json={
                    "eventId": event["id"],
                    "equipmentId": event["equipment_id"],
                    "equipmentName": event["equipment_name"],
                    "timestamp": datetime.now(timezone.utc).isoformat(),
                    **event["payload"],
                },
            )
            response.raise_for_status()
            logger.info(f"Apriso COUNTER published: event={event['id']} status={response.status_code}")

    async def test_connection(
        self, base_url: str, username: str = "", password: str = "", api_key: str = ""
    ) -> dict[str, Any]:
        if _is_stub(base_url):
            return {"success": True, "message": "Apriso is in stub/mock mode. Connection test simulated successfully."}
        try:
            headers = {"Content-Type": "application/json"}
            if api_key:
                headers["X-API-Key"] = api_key
            auth = (username, password) if username else None
            async with httpx.AsyncClient(
                base_url=base_url, headers=headers, auth=auth, timeout=10.0
            ) as client:
                await client.get("/api/v1/health")
            return {"success": True, "message": "Apriso connection successful"}
        except Exception as exc:
            return {"success": False, "message": f"Apriso connection failed: {exc}"}


apriso_service = AprisoService()
